content/doc/developer/FIPS-140/index.adoc

```diff
@@ -0,0 +1,46 @@
+[IMPORTANT]
+====
+FIPS-140 compliant code may result in less security as the approved standards can lag behind more secure modern standards.
+Your plugin and Jenkins code should not automatically follow the tips here, but may also need to make decisions at runtime whether or not to use a FIPS-140 compliant option.
+====
+
+Any sensitive information for accessing an external system that is persisted should be stored either as a link:https://javadoc.jenkins.io/hudson/util/Secret.html[`hudson.util.Secret`] or a link:https://javadoc.jenkins.io/plugin/credentials/com/cloudbees/plugins/credentials/Credentials.html[`com.cloudbees.plugins.credentials.Credentials`]
+
+Rely only on link:https://docs.oracle.com/en/java/javase/11/security/java-cryptography-architecture-jca-reference-guide.html#GUID-3E0744CE-6AC7-4A6D-A1F6-6C01199E6920[Java Cryptography Architecture (JCA)] or link:https://bouncycastle.org/java.html[BouncyCastle] via the link:https://plugins.jenkins.io/bouncycastle-api/[BouncyCastle API plugin] when performing cryptography.
+Do not depend on any third-party libraries or write your own code to perform encryption or specify any specific link:https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/security/Provider.html[`java.security.Provider`] when obtaining a concrete JCA implementation.
+Additionally, the subset of algorithms that are used should be restricted to the subset that is link:https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/3514[supported] by the link:https://downloads.bouncycastle.org/fips-java/BC-FJA-UserGuide-1.0.2.pdf[BouncyCastle FIPS] implementation.
+[NOTE]
+====
+FIPS-140 has no support for asymmetric encryption such as RSA.
+====
+
+If there is a tradeoff made between FIPS-140 support or a more secure option, the tradeoff should be made at runtime by checking link:https://javadoc.jenkins.io/jenkins/security/FIPS140.html#useCompliantAlgorithms()[`jenkins.security.FIPS140.useCompliantAlgorithms`].
+An example of changing behaviour can be found in link:https://github.com/jenkinsci/jenkins/pull/8483[this pull request].
+
+If Using a link:https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/security/SecureRandom.html[`java.security.SecureRandom`] it should be obtained via link:https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/security/SecureRandom.html#%3Cinit%3E()[`new SecureRandom()`] or link:https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/security/SecureRandom.html#getInstanceStrong()[`SecureRandom.getInstanceStrong()`].
+Do not use any of the methods or constructors that take an optional `seed`.
+
+[NOTE]
+====
+This guidance is applicable to all libraries that your plugin bundles, or uses from Jenkins core. 
+Jenkins core may provide some libraries that are not FIPS-140 compliant for some API calls, but are when called by Jenkins core itself.
```

Comment:
```suggestion
Jenkins core may provide some libraries that are not FIPS-140 compliant for some API calls, but are compliant when called by Jenkins core itself.
```
I am not sure if this is what you meant here, just wanted to make it concrete

Reply:
@julieheard I am trying to say just because the library is provided by Jenkins core, that it may not be safe to use, so should not be assumed as compliant.

```suggestion
Jenkins core may provide some libraries that are not FIPS-140 compliant for some API calls, but Jenkins itself may not utilize these specific methods.
```